autoprecompiles/src/lib.rs

```diff
@@ -0,0 +1,1213 @@
+                    .iter()
+                    .zip_eq(&pc_lookup.args)
+                    .for_each(|(instr_arg, pc_arg)| {
+                        let arg = AlgebraicExpression::Number(*instr_arg);
+                        match pc_arg {
+                            AlgebraicExpression::Reference(ref arg_ref) => {
+                                sub_map.insert(Column::from(arg_ref), arg);
+                            }
+                            AlgebraicExpression::BinaryOperation(_expr) => {
+                                local_constraints.push((arg - pc_arg.clone()).into());
+                            }
+                            AlgebraicExpression::UnaryOperation(_expr) => {
+                                local_constraints.push((arg - pc_arg.clone()).into());
+                            }
+                            _ => {}
+                        }
+                    });
+
+                machine.constraints.extend(local_constraints);
+
+                // Start the local substitutions with the constant columns
+                let mut local_subs = constant_subs.clone();
+
+                // Process all expressions in the machine
+                machine.visit_expressions_mut(
+                    &mut |expr| {
+                        *expr = {
+                            let mut expr = expr.clone();
+                            powdr::substitute_algebraic(&mut expr, &sub_map);
+                            powdr::substitute_algebraic(&mut expr, &sub_map_loadstore);
+                            let mut expr = simplify_expression(expr);
+                            global_idx =
+                                powdr::local_to_global(&mut expr, global_idx, &mut local_subs, i);
+                            expr
+                        };
+                        ControlFlow::Continue::<()>(())
+                    },
+                    VisitOrder::Pre,
+                );
+
+                result.constraints.extend(machine.constraints);
+                result.bus_interactions.extend(machine.bus_interactions);
+                col_subs.push(local_subs);
+
+                // after the first round of simplifying,
+                // we need to look for register memory bus interactions
+                // and replace the addr by the first argument of the instruction
+                for bus_int in &mut result.bus_interactions {
+                    if bus_int.id != MEMORY_BUS_ID {
+                        continue;
+                    }
+
+                    let addr_space = match bus_int.args[0] {
+                        AlgebraicExpression::Number(n) => n.to_integer().try_into_u32().unwrap(),
+                        _ => panic!(
+                            "Address space must be a constant but got {}",
+                            bus_int.args[0]
+                        ),
+                    };
+
+                    if addr_space != 1 {
+                        continue;
+                    }
```

Comment:
is it correct that we run this on the entire apc so far, as opposed to on the added interactions, or a single time on the whole thing at the end?